relay-general/src/store/clock_drift.rs

```diff
@@ -0,0 +1,263 @@
+#[derive(Clone, Copy, Debug)]
+struct HumanDuration(SignedDuration);
+
+impl fmt::Display for HumanDuration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let days = self.0.num_days();
+        if days.abs() == 1 {
+            write!(f, "{} day", days)
+        } else if days != 0 {
+            write!(f, "{} days", days)
+        } else if self.0.num_hours() != 0 {
+            write!(f, "{}h", self.0.num_hours())
+        } else {
+            write!(f, "{}s", self.0.num_seconds())
+        }
+    }
+}
+
+/// Corrects clock drift based on the sender's and receivers timestamps.
+///
+/// Clock drift correction applies to all timestamps in the event protocol. This includes especially
+/// the event's timestamp, breadcrumbs and spans.
+///
+/// There is a minimum clock drift of _55 minutes_ to compensate for network latency and small clock
+/// drift on the sender's machine, but allow to detect timezone differences. For differences lower
+/// than that, no correction is performed.
+///
+/// Clock drift is corrected in both ways:
+///
+/// - The drift is added to timestamps if the received time is after the send time. This indicates
+///   that the sender's clock was lagging behind. For instance, if an event was received with
+///   yesterday's timestamp, one day is added to all timestamps.
+///
+/// - The drift is subtracted from timestamps if the received time is before the send time. This
+///   indicates that the sender's clock was running ahead. For instance, if an event was received
+///   with tomorrow's timestamp, one day is subtracted from all timestamps.
+#[derive(Debug)]
+pub struct ClockDriftProcessor {
+    received_at: DateTime<Utc>,
+    correction: Option<ClockCorrection>,
+}
+
+impl ClockDriftProcessor {
+    /// Creates a new `ClockDriftProcessor`.
+    ///
+    /// If no `sent_at` timestamp is provided, then clock drift correction is disabled. The drift is
+    /// calculated from the signed difference between the receiver's and the sender's timestamp.
+    pub fn new(sent_at: Option<DateTime<Utc>>, received_at: DateTime<Utc>) -> Self {
+        let correction = sent_at.and_then(|sent_at| {
+            ClockCorrection::new(sent_at, received_at)
+                .at_least(SignedDuration::seconds(MINIMUM_CLOCK_DRIFT_SECS))
+        });
+
+        Self {
+            received_at,
+            correction,
+        }
+    }
+}
+
+impl Processor for ClockDriftProcessor {
+    fn process_event(
+        &mut self,
+        event: &mut Event,
+        _meta: &mut Meta,
+        state: &ProcessingState<'_>,
+    ) -> ProcessingResult {
+        event.process_child_values(self, state)?;
+
+        if let Some(correction) = self.correction {
+            let timestamp_meta = event.timestamp.meta_mut();
+            timestamp_meta.add_error(Error::with(ErrorKind::InvalidData, |e| {
+                let reason = format!(
+                    "clock drift: all timestamps adjusted by {}",
```

Comment:
Not a huge fan of producing error messages here, we might want to refactor _meta to emit reason *codes* (not related to this PR in particular, just for the future)

Reply:
The `ErrorKind` you saw is such a reason code, but I didn't want to touch Sentry for this now. If you like, we can follow up with a new error kind and then teach Sentry about it.